### PR TITLE
reworked _refund_cleanup_lines and test refund from claim

### PR DESCRIPTION
**Pull request by adamwojt**
https://hive.versada.eu/work_packages/11525

- _refund_cleanup_lines now uses index to add claim values, there is no need
to C/P Odoo's code to acheive this
- _refund_cleanup_lines now returns empty list for account.invoice.tax calls to it,
this is needed because origin tax lines are sum-up of all taxes and in case of claim
not all invoice lines are refunded. This also fixes wrong field name, see https://github.com/OCA/rma/issues/147

- [Original Odoo method](https://github.com/odoo/odoo/blob/91807404e7d5c314ea30926b44bd16e878417ec1/addons/account/models/account_invoice.py#L985)
- [Used here](https://github.com/odoo/odoo/blob/91807404e7d5c314ea30926b44bd16e878417ec1/addons/account/models/account_invoice.py#L1039)
- [OCA artwork](https://github.com/OCA/rma/blob/f0656eb39c55081eff52ad502a0a1deb648b0822/crm_claim_rma/models/account_invoice.py#L16)

### CODE REVIEW

File: crm_claim_rma/models/account_invoice.py [hunk 1]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
     claim_id = fields.Many2one('crm.claim', string='Claim')
 
+    @api.model
     def _refund_cleanup_lines(self, lines):
-        """ Override when from claim to update the quantity and link to the
-        claim line.
+        """ Override when from claim to update the quantity, link to the
+        claim line and return empty list for tax_line_ids.
         """
+        claim_lines_context = self._context['claim_line_ids']
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py]: 1. always try to use public methods/variables when they are available
```suggestion
        claim_lines_context = self.env.context['claim_line_ids']
```
2. Is `claim_line_ids` guaranteed to be in the context. I assume it is not based on this [code](https://github.com/versada/rma/blob/e0b02f63de1991ad6bbb0331752ffb76810eb2a6/crm_claim_rma/models/account_invoice.py#L22-L23) so you need to account for that

File: crm_claim_rma/models/account_invoice.py [hunk 2]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
 
-        # check if is an invoice_line and we are from a claim
-        if not (self.env.context.get('claim_line_ids') and lines and
-                lines[0]._name == 'account.invoice.line'):
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py:23]: Could you elaborate on why was this removed ?

File: crm_claim_rma/models/account_invoice.py [hunk 3]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
-        if not new_lines:
-            # TODO use custom states to show button of this wizard or
-            # not instead of raise an error
+        # We don't want tax_line_ids from origin invoice as new ones
+        # will be calculated for correct number of lines
+
+        if lines[0]._name == 'account.invoice.tax':
+            return []
+
+        claim_lines = self.env['claim.line'].browse(zip(*claim_lines_context)[1])
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py]: interesting use of `zip`, however in this case it really becomes quite unreadable. Use `zip` when you have more than 1 list and want to consequently have a joint result of both list items. In this case you join the lines and take just the first item, rendering other joins essentially useless. I think the original was good enough: `[l[1] for l in claim_lines_context]`

File: crm_claim_rma/models/account_invoice.py [hunk 4]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
+
+        if not claim_lines_wo_refund:
 
             raise exceptions.UserError(
-                _('A refund has already been created for this claim !')
+                _('A refund has already been created for all lines on this claim !')
             )
-        return [(0, 0, l) for l in new_lines]
+
+        # both mapped and _refund_cleanup_lines keep original sequence
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py]: I trust that you probably checked that, but don't rely on [it](https://github.com/odoo/odoo/blob/91807404e7d5c314ea30926b44bd16e878417ec1/odoo/models.py#L5016-L5017). Its even stated explicitly in odoo code that order is arbitrary in case of recordsets of the function `mapped`. Use a `dict` lookup instead for arbitrary order.

File: crm_claim_rma/models/account_invoice.py [hunk 5]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
+            }
+            for line in claim_lines_wo_refund]
+
+        for index, line in enumerate(refund_invoice_lines_data):
+            line[2].update(
+                quantity=claim_lines_data[index]['quantity'],
+                claim_line_id=claim_lines_data[index]['claim_line_id'],
+            )
+
+        return refund_invoice_lines_data
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py]: All of this can be done in an in place modification:
```python
print(list_of_dicts)
for one_dict in list_of_dicts:
   one_dict['value'] = 5
print(list_of_dicts) # this will be modified
```

File: crm_claim_rma/models/account_invoice.py [hunk 6]
```diff
@@ -13,47 +13,53 @@ class AccountInvoice(models.Model):
...
-
-                new_lines.append(clean_line)
-        if not new_lines:
-            # TODO use custom states to show button of this wizard or
-            # not instead of raise an error
+        # We don't want tax_line_ids from origin invoice as new ones
+        # will be calculated for correct number of lines
+
+        if lines[0]._name == 'account.invoice.tax':
+            return []
```
Comment (kkarolis) [inline crm_claim_rma/models/account_invoice.py:32]: seems like if we only operate on `account.invoice.line` this shouldn't be necessary ?

File: crm_claim_rma/tests/test_claim.py [hunk 7]
```diff
@@ -30,3 +30,37 @@ def test_create__with_claim_type(self):
         self.assertEqual(supplier_type, claim.claim_type)
         self.assertIsNotNone(claim.code)
         self.assertTrue(claim.code.startswith('RMA-V/'))
+
+    def test_refund_from_claim(self):
+        invoice = self.env.ref('l10n_generic_coa.demo_invoice_1')
```
Comment (kkarolis) [inline crm_claim_rma/tests/test_claim.py]: @naglis can you comment on why these tests work ? I was always lead to believe you can't actually use demo invoice data if you don't depend on the module.

**Review by kkarolis — CHANGES_REQUESTED** [review]
In forked repo's please always start from the main branch (`10.0` in this case). We maintain a separate branch (e.g. `10.0-versada`) so that we wouldn't need to wait for PR to be merged upstream.

File: crm_claim_rma/models/account_invoice.py (hunk 2 above)
Comment (adamwojt) [reply crm_claim_rma/models/account_invoice.py:23]: @kkarolis This method is used twice at the moment - [here](https://github.com/odoo/odoo/blob/91807404e7d5c314ea30926b44bd16e878417ec1/addons/account/models/account_invoice.py#L1066). Currently tax_line_ids are being appended to new invoice. In case of full refund (odoo default), this is ok. In case of claim that has only few lines - taxes are incorrect, therefore we are returning empty list for `account.invoice.tax` .

File: crm_claim_rma/models/account_invoice.py (hunk 6 above)
Comment (adamwojt) [reply crm_claim_rma/models/account_invoice.py:32]: As above.

File: crm_claim_rma/tests/test_claim.py (hunk 7 above)
Comment (naglis) [reply crm_claim_rma/tests/test_claim.py]: @kkarolis why me :slightly_smiling_face: ?

IIUC, the `l10n_generic_coa` addon is [installed implicitly](https://github.com/odoo/odoo/blob/62d7eaf865ca31574b8efb73d3cfb7836ed60da3/addons/account/__init__.py#L28) if no other country `l10n_*` addon is found (based on company country) via a [post-init hook](https://github.com/odoo/odoo/blob/62d7eaf865ca31574b8efb73d3cfb7836ed60da3/addons/account/__manifest__.py#L79) when `account` addon is installed. This is nasty, but depending on `l10n_generic_coa` is not an option to allow the addon to be used with other countries, so I guess there should be a check in the test to skip it if `l10n_generic_coa` is not loaded.

Comment (adamwojt) [reply crm_claim_rma/tests/test_claim.py]: I will use other ref not to complicate things.

Comment (kkarolis) [reply crm_claim_rma/tests/test_claim.py]: @naglis IIRC we had a small discussion similar to this in some other project, forgive me if i'm wrong :)

Comment (kkarolis) [reply crm_claim_rma/tests/test_claim.py]: @adamwojt using other ref is not actually a solution if its from `l10n_generic_coa`. The problem I had with this is that implicit install and I remember in some tests having issues with that. Anyway, I suggest to leave it as is.

Comment (adamwojt) [reply crm_claim_rma/tests/test_claim.py]: @kkarolis I used sale demo ref and created invoice from there.

Comment (naglis) [reply crm_claim_rma/tests/test_claim.py]: @kkarolis np, but I can't remember. Sorry, if that's the case.

**Comment by adamwojt** [comment]
Hey @kkarolis I squashed everything to one commit along with changes we discussed in PR and cherry-pick added it to 10.0 branch . 
I am sorting `claim_lines_wo_refund` by `invoice_line_id`. 
Then the order for `invoice_lines` is [default](https://github.com/odoo/odoo/blob/91807404e7d5c314ea30926b44bd16e878417ec1/addons/account/models/account_invoice.py#L1185). Not sure if I should sort by `id` there - sequence seem to be increasing same as id ?

**Review by kkarolis — APPROVED** [review]

**Review by naglis — APPROVED** [review]
